src/client/automation/playback/scroll.js

```diff
@@ -143,24 +152,60 @@ export default class ScrollAutomation {
-
-        var childPoint = {
+    _getScrollPosition (parentDimensions, childDimensions, offsetX, offsetY) {
+        const childPoint = {
             x: childDimensions.left - parentDimensions.left + parentDimensions.scroll.left +
                childDimensions.border.left + offsetX,
             y: childDimensions.top - parentDimensions.top + parentDimensions.scroll.top +
                childDimensions.border.top + offsetY
         };
 
-        var scrollToFullView = this._getScrollToFullChildView(parentDimensions, childDimensions);
-        var scrollToPoint    = this._getScrollToPoint(parentDimensions, childPoint);
+        const scrollToPoint    = this._getScrollToPoint(parentDimensions, childPoint);
+        const scrollToFullView = this._getScrollToFullChildView(parentDimensions, childDimensions);
 
-        var left = scrollToFullView.left === null ? scrollToPoint.left : scrollToFullView.left;
-        var top  = scrollToFullView.top === null ? scrollToPoint.top : scrollToFullView.top;
+        const left = Math.max(scrollToFullView.left === null ? scrollToPoint.left : scrollToFullView.left, 0);
+        const top  = Math.max(scrollToFullView.top === null ? scrollToPoint.top : scrollToFullView.top, 0);
 
-        return this._setScroll(parent, { left, top });
+        return { left, top };
+    }
+
+    _getChildPointAfterScroll (parentDimensions, childDimensions, left, top) {
+        const x = childDimensions.left + parentDimensions.scroll.left - left +
+                  Math.round(childDimensions.width / 2);
+        const y = childDimensions.top + parentDimensions.scroll.top - top +
+                  Math.round(childDimensions.height / 2);
+
+        return { x, y };
+    }
+
+    _scrollToChild (parent, child, { offsetX, offsetY }) {
+        const parentDimensions = positionUtils.getClientDimensions(parent);
+        const childDimensions  = positionUtils.getClientDimensions(child);
+        const windowWidth      = styleUtils.getInnerWidth(window);
+        const windowHeight     = styleUtils.getInnerHeight(window);
+        let scrollPos          = {};
+        let needScroll         = true;
+
+        while (needScroll) {
+            scrollPos = this._getScrollPosition(parentDimensions, childDimensions, offsetX, offsetY);
+
+            const { x, y }         = this._getChildPointAfterScroll(parentDimensions, childDimensions, scrollPos.left, scrollPos.top);
+            const isTargetObscured = this._isTargetElementObscuredInPoint(x, y);
+
+            this.maxScrollMarginLeft += SCROLL_MARGIN_INCREASE_STEP;
+
+            if (this.maxScrollMarginLeft >= windowWidth) {
+                this.maxScrollMarginLeft = DEFAULT_MAX_SCROLL_MARGIN;
+                this.maxScrollMarginTop += SCROLL_MARGIN_INCREASE_STEP;
+            }
+
+            needScroll = isTargetObscured && this.maxScrollMarginTop < windowHeight;
+        }
+
+        this.maxScrollMarginLeft = DEFAULT_MAX_SCROLL_MARGIN;
```

Comment:
Can we use one `this.maxScrollMargin = { left: ..., top ...}` property instead of `this.maxScrollMarginLeft` and `this.maxScrollMarginTop`?